File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        # for a given container, it is the height of the smaller side * 
        max_water = 0
        l, r = 0, len(heights) - 1
        while l < r:
            max_water = max((r - l) * min(heights[r], heights[l]), max_water)
            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1
        return max_water

File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int area;
        int maxArea = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == 1) {
                    area = dfs(i, j, grid);
                    maxArea = max(area, maxArea);
                }
            }
        }
        return maxArea;
    }

    int dfs(int x, int y, vector<vector<int>>& grid) {
        int area = 1;
        if (grid[x][y] == 0) {
            return 0;
        } else {
            grid[x][y] = 0;
            if (x + 1 < grid.size()) {
                area += dfs(x + 1, y, grid);
            }
            if (x - 1 >= 0) {
                area += dfs(x - 1, y, grid);
            }
            if (y + 1 < grid[0].size()) {
                area += dfs(x, y + 1, grid);
            }
            if (y - 1 >= 0) {
                area += dfs(x, y - 1, grid);
            }

            return area;
        }

    }
};
